gui: add warnings for descendant & conflicting transactions that will be dropped

jp1ac4 opened the pull request:
This is for #903.

It adds two warnings to the GUI:
- when creating a new RBF if there are any descendant transactions of the transaction to be replaced:
![image](https://github.com/wizardsardine/liana/assets/121959000/b3d91253-fa5f-4c4c-b3f7-1e22d56ff0ba)

- when broadcasting a transaction if there are any conflicting transactions
![image](https://github.com/wizardsardine/liana/assets/121959000/c3c8d932-e2d9-4982-9f5d-c7cf15d53370)

These warnings are generated in the GUI using coins data from the DB and so will not appear if the DB coins have not yet been updated. Would resolving https://github.com/wizardsardine/liana/issues/887 help ensure the coins are updated before running these checks?

darosior commented on gui/src/app/view/psbt.rs:
```diff
@@ -151,15 +159,50 @@ pub fn broadcast_action<'a>(warning: Option<&Error>, saved: bool) -> Element<'a,
+                        Row::new()
+                            .push(icon::warning_icon())
+                            .push(text(format!(
+                                "WARNING: The following transaction{} spending one or more \
+                                inputs from the transaction to be broadcast and will be \
+                                dropped, along with any other transactions that depend on {}:",
```
Let's have a higher level first sentence before the technical transaction dependency description? I'm thinking something along the lines of "WARNING: replacing this transaction would invalidate some later payments. [...]".

darosior commented on gui/src/app/view/transactions.rs:
```diff
@@ -169,15 +174,52 @@ pub fn create_rbf_modal<'a>(
+                    Row::new()
+                        .push(icon::warning_icon())
+                        .push(text(format!(
+                            "WARNING: The following transaction{} spending one or \
+                            more outputs from the transaction to be replaced and will \
+                            be dropped when the replacement is broadcast, along with \
+                            any other transactions that depend on {}:",
```
Same here.

darosior reviewed (commented):
Approach ACK.

> These warnings are generated in the GUI using coins data from the DB and so will not appear if the DB coins have not yet been updated. Would resolving https://github.com/wizardsardine/liana/issues/887 help ensure the coins are updated before running these checks?

In most cases yes. It wouldn't cover it if it was just broadcast from another wallet (one such realistic scenario would be in the case of a multi-user wallet such as a multisig). But i think that's fine.

jp1ac4 commented:
I've made some changes to the wording as suggested. I also added the option to copy the transaction ID with a copy icon.

Here are some screenshots for the broadcast warning:

- a single transaction:
![image](https://github.com/wizardsardine/liana/assets/121959000/8c454767-3567-4d6e-b2b4-493e0c790381)

- more than one transaction:
![image](https://github.com/wizardsardine/liana/assets/121959000/92ec834b-9118-46e2-8682-e9d89d3177b0)

For cancelling a transaction:
- a single transaction:
![image](https://github.com/wizardsardine/liana/assets/121959000/25e248fb-64b7-47c1-9e67-6c71faf7b2d9)

- more than one transaction:
![image](https://github.com/wizardsardine/liana/assets/121959000/11c5cd4f-5339-427a-9642-6faf74298599)

And for bumping the fee:
![image](https://github.com/wizardsardine/liana/assets/121959000/ff2c1177-700e-4db5-a8a7-ab3324ebde41)

I also reworded the existing "cancel transaction" message to mention sending coins "back to your wallet" instead of "back to us".

darosior commented:
That looks very good. Thank you for all the screenshots.

edouardparis commented on gui/src/app/state/transactions.rs:
```diff
@@ -126,8 +127,24 @@ impl State for TransactionsPanel {
                                 .unwrap();
-                            let modal =
-                                CreateRbfModal::new(tx.tx.txid(), is_cancel, prev_feerate_vb);
+                            let descendant_txids: HashSet<_> = cache
```
I would like we do not use the cache coins for this usage. cache is generally used to display something like the previous balance, while the new balance is calculated in order to not display zero values.

I think it would be good instead, to have CreateRbfModal::new() -> (modal, command)
and give the command result to modal.update so it updates its conflicting txids.

What do you think ?

jp1ac4 replied:
That sounds good thanks, I'll make this change. After we implement https://github.com/wizardsardine/liana/issues/677, then we'll be able to retrieve just the spending coins in this call.

jp1ac4 replied:
I've now made the change to use `daemon.list_coins` and pass the result to another message.

In the end, I didn't pass the result to `modal.update` because when I tried this, the modal would open without any warnings and then update shortly afterwards, and I felt it was better that the warnings would be there as soon as the modal opened. So the modal is created using the result of `list_coins`.

I made a similar change for the set of conflicting transactions when broadcasting a transaction.

jp1ac4 commented:
As well as the changes discussed above in https://github.com/wizardsardine/liana/pull/930#discussion_r1473035686, I've also removed the `prev_feerate_vb` parameter from `CreateRbfModal::new()` as this value can now be obtained from the `tx` parameter.

In order for the new `BroadcastModal` message handler to access `Error` instead of `&Error`, I changed the `match` statement to use `message` instead of `&message`.

edouardparis commented:
ACK fde28bf571e4ba791617e62f0c124138cd58e1fe

edouardparis reviewed (approved):
ACK fde28bf571e4ba791617e62f0c124138cd58e1fe
